File: libs/api/src/main/java/com/example/api/ApiCallAdapter.java
package com.example.api;

import android.support.annotation.NonNull;

import java.lang.reflect.Type;

import io.reactivex.Observable;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.Callback;
import retrofit2.Response;

class ApiCallAdapter<R> implements CallAdapter<R, Observable<ApiResponse<R>>> {
  private final Type typeOfResponse;

  ApiCallAdapter(Type typeOfResponse) {
    this.typeOfResponse = typeOfResponse;
  }

  @Override
  public Type responseType() {
    return typeOfResponse;
  }

  @Override
  public Observable<ApiResponse<R>> adapt(@NonNull Call<R> call) {
    return Observable.create(emitter ->
      call.enqueue(new Callback<R>() {
        @Override
        public void onResponse(@NonNull Call<R> call, @NonNull Response<R> response) {
          emitter.onNext(new ApiResponse<>(response));
        }

        @Override
        public void onFailure(@NonNull Call<R> call, @NonNull Throwable throwable) {
          emitter.onError(throwable);
        }
      })
    );
  }
}

File: libs/api/src/main/kotlin/com/example/api/ApiCallAdapter.kt
package com.example.api

import io.reactivex.Observable
import retrofit2.Call
import retrofit2.CallAdapter
import retrofit2.Callback
import retrofit2.Response
import java.lang.reflect.Type

internal class ApiCallAdapter<R>(private val typeOfResponse: Type) : CallAdapter<R, Observable<ApiResponse<R>>> {
  override fun responseType(): Type {
    return typeOfResponse
  }

  override fun adapt(call: Call<R>): Observable<ApiResponse<R>> {
    return Observable.create { emitter ->
      call.enqueue(object : Callback<R> {
        override fun onResponse(call: Call<R>, response: Response<R>) {
          emitter.onNext(ApiResponse(response))
        }

        override fun onFailure(call: Call<R>, throwable: Throwable) {
          emitter.onError(throwable)
        }
      })
    }
  }
}
